plugins/drude/tests/TestDrudeNoseHoover.h

```diff
@@ -0,0 +1,605 @@
+        system.addParticle(1.0);  // H2
+        system.addParticle(0.0);  // M
+        nonbonded->addParticle(1.71636, 0.318395, 0.21094*4.184);
+        nonbonded->addParticle(-1.71636, 1, 0);
+        nonbonded->addParticle(0.55733, 1, 0);
+        nonbonded->addParticle(0.55733, 1, 0);
+        nonbonded->addParticle(-1.11466, 1, 0);
+        for (int j = 0; j < 5; j++)
+            for (int k = 0; k < j; k++)
+                nonbonded->addException(startIndex+j, startIndex+k, 0, 1, 0);
+        system.addConstraint(startIndex, startIndex+2, 0.09572);
+        system.addConstraint(startIndex, startIndex+3, 0.09572);
+        system.addConstraint(startIndex+2, startIndex+3, 0.15139);
+        system.setVirtualSite(startIndex+4, new ThreeParticleAverageSite(startIndex, startIndex+2, startIndex+3, 0.786646558, 0.106676721, 0.106676721));
+        drude->addParticle(startIndex+1, startIndex, -1, -1, -1, -1.71636, polarizability, 1, 1);
+    }
+    for (int i = 0; i < gridSize; i++) {
+        for (int j = 0; j < gridSize; j++) {
+            for (int k = 0; k < gridSize; k++) {
+                Vec3 pos(i*spacing, j*spacing, k*spacing);
+                positions.push_back(pos);
+                positions.push_back(pos);
+                positions.push_back(pos+Vec3(0.09572, 0, 0));
+                positions.push_back(pos+Vec3(-0.023999, 0.092663, 0));
+                positions.push_back(pos);
+            }
+        }
+    }
+}
+
+void testWaterBox(Platform& platform) {
+    // Create a box of SWM4-NDP water molecules.  This involves constraints, virtual sites,
+    // and Drude particles.
+    System system;
+    const int gridSize = 3;
+    vector<Vec3> positions;
+    double polarizability = ONE_4PI_EPS0*1.71636*1.71636/(100000*4.184);
+
+    build_waterbox(system, gridSize, polarizability, positions);
+
+    const int numMolecules = gridSize*gridSize*gridSize;
+    int numStandardDof = 3*3*numMolecules - system.getNumConstraints();
+    int numDrudeDof = 3*numMolecules;
+    int numDof = numStandardDof+numDrudeDof;
+    const double temperature = 300.0;
+    const double temperatureDrude = 10.0;
+
+    // Simulate it and check the temperature.
+    int chainLength = 4;
+    int numMTS = 4;
+    int numYS = 5;
+    double frequency = 800.0;
+    double frequencyDrude = 2000.0;
+    int randomSeed = 100;
+    DrudeNoseHooverIntegrator integ(temperature, frequency, 
+                                    temperatureDrude, frequencyDrude, 0.0005, 
+                                    chainLength, numMTS, numYS);
+    Context context(system, integ, platform);
+    context.setPositions(positions);
+    context.setVelocitiesToTemperature(temperature, randomSeed);
+    std::vector<Vec3> velocities = context.getState(State::Velocities).getVelocities();
+    for (int i = 0; i < numMolecules; i++){
+        Vec3 noize;
+        for (int j = 0; j < 3; j++){
+            noize[j] = float(((i+18311)*(j+18253) * 313419097822414) % 18313) / float(18313);
```

Comment:
I take it this is a random number generator? 😀  We do have a proper RNG available for use in tests:

```c++
#include "sfmt/SFMT.h"
...
// Create and initialize an RNG instance
OpenMM_SFMT::SFMT sfmt;
init_gen_rand(0, sfmt);
...
noise[j] = genrand_real2(sfmt); // Uniformly distributed between 0 and 1
```

Reply:
:-) We were not sure if the built-in RNG is 100% platform-independent. But if you don't mind we will leave it as is - this would spare us to rewrite the check of positions and velocities below :-)